File: app/pipelines/news.py
"""News by keyword pipeline using Google News RSS."""

import logging
from urllib.parse import quote_plus

import feedparser

from app.models import db, AlertHistory, TrackedEntity

logger = logging.getLogger(__name__)

GOOGLE_NEWS_RSS = "https://news.google.com/rss/search?q={query}&hl=en"


def check_news(entity_names, user_id):
    """Check for news articles matching tracked keywords.

    Returns a list of new result dicts.
    """
    all_new = []

    for name in entity_names:
        entity = TrackedEntity.query.filter_by(
            user_id=user_id, entity_name=name
        ).join(TrackedEntity.category).filter_by(slug="news").first()

        if not entity:
            continue

        try:
            results = _fetch_news(name)
        except Exception:
            logger.exception("Failed to fetch news for %s", name)
            continue

        seen_keys = _get_seen_keys(user_id, entity.id)

        for item in results:
            key = _make_key(item)
            if key in seen_keys:
                continue

            alert = AlertHistory(
                user_id=user_id,
                tracked_entity_id=entity.id,
            )
            alert.set_content(item)
            db.session.add(alert)
            all_new.append(item)

    db.session.commit()
    return all_new


def _fetch_news(keyword):
    """Fetch recent news articles from Google News RSS for a keyword."""
    url = GOOGLE_NEWS_RSS.format(query=quote_plus(keyword))
    feed = feedparser.parse(url)

    results = []
    for entry in feed.entries[:10]:
        results.append({
            "type": "news_article",
            "keyword": keyword,
            "title": entry.get("title", ""),
            "source": entry.get("source", {}).get("title", ""),
            "published": entry.get("published", ""),
            "link": entry.get("link", ""),
        })
    return results


def _make_key(item):
    """Create a dedup key from a result item."""
    return f"{item['keyword']}|{item['title']}"


def _get_seen_keys(user_id, entity_id):
    """Get set of already-seen dedup keys for this user+entity."""
    alerts = AlertHistory.query.filter_by(
        user_id=user_id, tracked_entity_id=entity_id
    ).all()
    keys = set()
    for a in alerts:
        content = a.get_content()
        keys.add(_make_key(content))
    return keys

File: packages/shared/src/pipelines/news.ts
import { PipelineContext, PipelineResult } from './types';

const GOOGLE_NEWS_RSS = 'https://news.google.com/rss/search';

export async function checkNews(ctx: PipelineContext): Promise<PipelineResult[]> {
  const results: PipelineResult[] = [];

  for (const entity of ctx.entities) {
    const keyword = entity.entity_name;

    try {
      const articles = await fetchNewsArticles(keyword);
      for (const article of articles) {
        const dedupKey = `${keyword}|${article.title}`;

        results.push({
          entity_name: keyword,
          tracked_entity_id: entity.id,
          dedup_key: dedupKey,
          event_date: article.pubDate || undefined,
          content: {
            type: 'news',
            title: article.title,
            url: article.link,
            source: article.source,
            published: article.pubDate,
          },
        });
      }
    } catch (err) {
      console.error(`News pipeline error for "${keyword}":`, err);
    }
  }

  return results;
}

interface NewsArticle {
  title: string;
  link: string;
  source: string;
  pubDate: string;
}

async function fetchNewsArticles(keyword: string): Promise<NewsArticle[]> {
  const url = `${GOOGLE_NEWS_RSS}?q=${encodeURIComponent(keyword)}&hl=en-US&gl=US&ceid=US:en`;

  const res = await fetch(url);
  if (!res.ok) return [];

  const xml = await res.text();
  return parseRSS(xml).slice(0, 10);
}

function parseRSS(xml: string): NewsArticle[] {
  const items: NewsArticle[] = [];
  const itemRegex = /<item>([\s\S]*?)<\/item>/g;
  let match;

  while ((match = itemRegex.exec(xml)) !== null) {
    const itemXml = match[1];
    const title = extractTag(itemXml, 'title');
    const link = extractTag(itemXml, 'link');
    const source = extractTag(itemXml, 'source');
    const pubDate = extractTag(itemXml, 'pubDate');

    if (title && link) {
      items.push({ title, link, source: source || 'Unknown', pubDate: pubDate || '' });
    }
  }

  return items;
}

function extractTag(xml: string, tag: string): string {
  const regex = new RegExp(`<${tag}[^>]*>(?:<!\\[CDATA\\[)?(.*?)(?:\\]\\]>)?</${tag}>`, 's');
  const match = regex.exec(xml);
  return match?.[1]?.trim() || '';
}
